### PR TITLE
feat(cli): add hashed assets for standard exports

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

@blainekasten has reported an issue with Metro websites where the assets don't perform browser cache invalidation because the production names don't contain content hashes. We use a special hashing system on native that isn't supported in `react-native-web` (`setCustomSourceTransformer`) so we need a quick workaround to inject content hashes in production. This PR adds the hash of the first asset to the `name` property so we have hashing for basic cases. Instead of using the traditional `<name>.<hash>@1x.<ext>` we use `<name>_<hash>` to support Android res folders, this is due to the limitation on Metro asset plugins which doesn't provide any platform information to the plugin.

We'll need to come back through and clean up the entire asset pipeline to support all our various cases in the future.

# Test Plan

Link `expo-asset`, `@expo/metro-config`, and `@expo/cli` to a new project and build production ios and android apps to ensure the assets are pulled in and used hashed names. Ensure the same with web.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/@expo/cli/CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 ### 🎉 New features
 
+- Add content hashing for assets that aren't hosted with EAS Updates. ([#25774](https://github.com/expo/expo/pull/25774) by [@EvanBacon](https://github.com/EvanBacon))
```
Comment (wschurman) [inline packages/@expo/cli/CHANGELOG.md:14]:
```suggestion
- Add content hashing for assets that aren't set by expo-updates. ([#25774](https://github.com/expo/expo/pull/25774) by [@EvanBacon](https://github.com/EvanBacon))
```

**Review by wschurman — COMMENTED** [review]
I think @ide was looking into this in https://github.com/expo/expo/pull/25614. Might be worth syncing with him.

**Comment by ide** [comment]
https://github.com/expo/expo/pull/25614 removes the MD5 checksum just to simplify the code, especially since update assets are local assets (not remote) from the perspective of expo-asset and also are already checksummed by expo-updates using SHA-1. I believe these two PRs are independent of each other.

For this PR: is there a way to achieve the same goals while keeping Android and iOS asset names intact? I feel worried about renaming assets, especially the documented coupling between Metro and expo-asset and also unanticipated effects of people/libraries relying on the current file names. Do other reviewers share this concern or is it too theoretical?

As a suggestion, is it possible to add a hash as a query parameter in the JS bundle? For example, a reference to `./icon.png` would become `./icon.png?_h={computeContentAddress(asset)}` where `computeContentAddress` could be `base64url(sha1(assetContent)).substring(0, 10)` (SHA-1 is preferable over MD5, base64url is more compact than hex, and 10 characters usually gives good enough entropy for cache invalidation purposes). The assets themselves could retain their original names (icon.png still gets served as icon.png), it's just the URLs that reference them that would be content-addressed.

Alternatively, if the issue is that the static file server responds with long-lived cache-control headers, then I think we'd need to rewrite the file names. But even in that case, would it be feasible to do this just for web?

**Comment by EvanBacon** [comment]
@ide I'll need to fork two parts of Metro but I think it's worth it to make this change web-only and therefore safer as we have extensive testing for web bundling behavior.

**Pull request closed without merging**